processing/src/main/java/io/druid/query/search/search/AutoStrategy.java

```diff
@@ -0,0 +1,56 @@
+package io.druid.query.search.search;
+
+import com.metamx.emitter.EmittingLogger;
+import io.druid.collections.bitmap.ImmutableBitmap;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.Segment;
+
+import java.util.List;
+
+public class AutoStrategy extends SearchStrategy
+{
+  public static final String NAME = "auto";
+
+  private static final EmittingLogger log = new EmittingLogger(AutoStrategy.class);
+
+  public AutoStrategy(SearchQuery query)
+  {
+    super(query);
+  }
+
+  @Override
+  public List<SearchQueryExecutor> getExecutionPlan(SearchQuery query, Segment segment)
+  {
+    final QueryableIndex index = segment.asQueryableIndex();
+
+    if (index != null) {
+      final ImmutableBitmap timeFilteredBitmap = IndexOnlyStrategy.makeTimeFilteredBitmap(index,
+                                                                                          segment,
+                                                                                          filter,
+                                                                                          interval);
+      final Iterable<DimensionSpec> dimsToSearch = getDimsToSearch(index.getAvailableDimensions(),
+                                                                   query.getDimensions());
+
+      // Choose a search query execution strategy depending on the query.
+      // Index-only strategy is selected when
+      // 1) there is no filter,
+      // 2) the total cardinality is very low, or
+      // 3) the filter has a very low selectivity.
```

Comment:
hmm, I wonder if we can do a better heuristic here. Instead of checking for "low cardinality" and "low selectivity" does it make sense to use cost functions like:

- cursor-based is `numRowsInSegment * filterSelectivity * costOfCheckingSingleValue`
- index-based with a filter is `dimensionCardinality * (costOfCheckingSingleValue + searchQuerySelectivity * costOfFilterIntersection)`
- index-based without a filter is `dimensionCardinality * costOfCheckingSingleValue`

And choose the strategy with the lowest cost. Unfortunately we don't know what searchQuerySelectivity is in advance, so I think we'd have to just use some constant.

One behavior difference is that if dimensionCardinality is equal to numRowsInSegment, like a unique key, then cost functions like that would choose the cursor-based strategy regardless of filter selectivity (since `dimensionCardinality >= numRowsInSegment * filterSelectivity` for all values of filterSelectivity). The current code would still use the index-based approach if filter selectivity is lower than the threshold. I'd guess that using cursor-based is actually better, though.

We can also skip computing timeFilteredBitmap if cursor-based already beats index-based (or comes close) before applying the `* filterSelectivity` term. Based on your benchmarks, it looks like this can save substantial time.

@jihoonson what do you think?